### PR TITLE
feat(ivy): Add support for Ivy version scheme

### LINKED ISSUE #3467: Support Ivy version scheme
Similar to Maven: https://ant.apache.org/ivy/history/2.3.0/ivyfile/dependency.html#revision

### PR DESCRIPTION
**Pull request by zharinov**
Closes #3467

### CODE REVIEW

File: lib/versioning/ivy/index.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  getMinor,
+  getPatch,
+  isGreaterThan,
+  isSingleVersion,
+  isStable,
+  matches: mavenMatches,
+  maxSatisfyingVersion,
+  minSatisfyingVersion,
+  getNewValue,
+  sortVersions,
```
Comment (zharinov) [inline lib/versioning/ivy/index.js]: Not sure about latest 4 functions reuse

**Comment by rarkins** [comment]
@zharinov can you separate the maven refactoring into a separate PR that we merge first?

File: lib/versioning/ivy/index.js (hunk 1 above)
Comment (rarkins) [reply lib/versioning/ivy/index.js]: I thought we wouldn't be able to directly reuse any of the functions due to the change in syntax? e.g.

```
[1.0,2.0[ matches all versions greater or equal to 1.0 and lower than 2.0
]1.0,2.0] matches all versions greater than 1.0 and lower or equal to 2.0
]1.0,2.0[ matches all versions greater than 1.0 and lower than 2.0
]1.0,) matches all versions greater than 1.0
(,2.0[ matches all versions lower than 2.0
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/versioning/ivy/parse.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    const value = str.replace(SUBREV_REGEX, '');
+    if (isSingleVersion(value)) {
+      return {
+        type: REV_TYPE_SUBREV,
+        value,
+      };
+    }
+  }
+
+  const range = parseRange(str.replace(/^]/, '(').replace(/\[$/, ')'));
```
Comment (zharinov) [inline lib/versioning/ivy/parse.js]: @rarkins Here i'm extending the range syntax, so it can be reused

Comment (rarkins) [reply lib/versioning/ivy/parse.js]: Sounds OK if you found a way to map them back and forth

Comment (zharinov) [reply lib/versioning/ivy/parse.js]: You mean, to preserve one or another style after extending? Yeah, that requires more than just couple replacings

**Comment by rarkins** [comment]
@zharinov can you add tests and remove WIP label when ready for merge?

**Comment by zharinov** [comment]
For now we have single implementation for ranges which covers usual `maven` and `ivy` use-cases. I propose to split this logic only if any problems will be found with it.

**Pull request merged**